### PR TITLE
nixos/mautrix-whatsapp: init

### PR DESCRIPTION
**Pull request by frederictobiasc**
Hi,

there already exists two merge requests regarding init of mautrix-whatsapp:

- https://github.com/NixOS/nixpkgs/pull/176025 seems abandoned since the original author does not reply anymore.
- https://github.com/NixOS/nixpkgs/pull/187416 configuration does not match current configuration of mautrix-whatsapp regarding logging.

I don't know what the canonical way of dealing with this situation is, but I would like to get a mautrix-whatsapp module introduced. That's why I did my best to pull together all suggestions from those two pull requests and also update them to be compatible with the current version of mautrix-whatsapp. Major changes were in the log configuration.

I'm currently running this configuration successfully on my server.

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [X] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by frederictobiasc** [comment]
@NickCao ping

File: nixos/modules/services/matrix/mautrix-whatsapp.nix [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        optional config.services.matrix-synapse.enable "matrix-synapse.service"
+      '';
+      description = lib.mdDoc ''
+        List of Systemd services to require and wait for when starting the application service.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    services.mautrix-whatsapp.settings = {
```
Comment (NickCao) [inline nixos/modules/services/matrix/mautrix-whatsapp.nix:81]: Can't these be set in the default section above? (which is done in the mautrix-telegram module)

Comment (frederictobiasc) [reply nixos/modules/services/matrix/mautrix-whatsapp.nix:81]: I encontered an [error with nix docs](https://github.com/NixOS/nixpkgs/actions/runs/5743104020/job/15566635874) when referencing another module from the default attribute set.
That's when I started doing some research on that topic. I found [this section in nixos manual](https://nixos.org/manual/nixos/stable/#sec-settings-nix-representable) and following that approach moving the default values to the config attribute set.

Comment (NickCao) [reply nixos/modules/services/matrix/mautrix-whatsapp.nix:81]: > If you insist, I could moving only the homeserver.domain attribute to settings and leave the rest in `default`.

This sounds like a better solution, at least it reduces the amount of `lib.mkDefault`

Comment (frederictobiasc) [reply nixos/modules/services/matrix/mautrix-whatsapp.nix:120]: Done.
But shouldn't the manual be updated then also?
I read Example 32 from the [NixOS manual](https://nixos.org/manual/nixos/stable/#sec-settings-nix-representable) as default values should be defined using `lib.mkDefault`.

Comment (NickCao) [reply nixos/modules/services/matrix/mautrix-whatsapp.nix:120]: AFIAK these two approaches should be equivalent in action. Just personally prefer the better looking one.

**Review by NickCao — APPROVED** [review]

File: nixos/modules/services/matrix/mautrix-whatsapp.nix (hunk 1 above)
Comment (frederictobiasc) [reply nixos/modules/services/matrix/mautrix-whatsapp.nix:120]: Thanks for the clarification! :)

**Comment by Luflosi** [comment]
It would have been nice if you had posted a comment in the other PRs so that everyone there would have gotten a notification. The only way I noticed this PR is that I rebuilt my system and the options introduced by this PR collided with the ones that I defined.

I think you made a small mistake by defining `appservicePort`. If a user changes the port by setting `appservice.port`, then `appservice.address` will still contain `http://localhost:29318`. I don't know what happens with this discrepancy but I think this should not happen. And it did not happen in a previous commit in this PR.

**Comment by Luflosi** [comment]
I'll create another PR soon™️ to dress this and a few other small things I found.

Thank you for finally getting this module into NixOS!

**Comment by Vskilet** [comment]
Thanks a lot for your work and sorry for my slow replies !

Someone has test https://gitlab.com/coffeetables/nix-matrix-appservices/-/tree/main ? It seems that would be nice to integrate https://github.com/NixOS/nixpkgs/pull/114419

Have great day

**Comment by eclairevoyant** [comment]
@frederictobiasc I don't see your name in the all-maintainers list. Can you add this in another PR, following these guidelines: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md

**Comment by frederictobiasc** [comment]
@eclairevoyant Hi and please excuse the delay. Thank you for spotting this. I created a PR: https://github.com/NixOS/nixpkgs/pull/252615

**Pull request merged**